File: coinbase/connect4.py
class Connect4:
    def __init__(self, n: int, m: int):
        self.rows = n
        self.columns = m
        self.board = [['_' for _ in range(self.columns)] for _ in range(self.rows)]
        self.currentPlayer = 'X'

    def __count_direction(self, row: int, col: int, delta_row: int, delta_col: int) -> int:
        count = 0
        r = row + delta_row
        c = col + delta_col
        while r >= 0 and r < self.rows and c >=0 and c < self.columns and self.board[r][c] == self.currentPlayer:
            count+=1
            r += delta_row
            c += delta_col
        return count
    
    def __check_win(self, row: int, col: int) -> bool:
        if 1 + self.__count_direction(row, col, 0, -1) + self.__count_direction(row, col, 0, 1) >= 4:
            return True
        
        if 1 + self.__count_direction(row, col, -1, 0) + self.__count_direction(row, col, 1, 0) >= 4:
            return True
        
        if 1 + self.__count_direction(row, col, -1, 1) + self.__count_direction(row, col, 1, -1) >= 4:
            return True
        
        if 1 + self.__count_direction(row, col, -1, -1) + self.__count_direction(row, col, 1, 1) >= 4:
            return True

        return False
    
    def __get_available_row(self, col) -> int:
        for i in range(self.rows - 1, -1, -1):
            if self.board[i][col] == '_':
                return i
    
    def display_board(self):
        for row in self.board:
            for cell in row:
                print(f"{cell:>2}", end=" ")
            print()
        
        for i in range(self.columns):
            print(f"{i:>2}", end=" ")
        print()

    def reset_board(self):
        self.board = [['_' for _ in range(self.columns)] for _ in range(self.rows)]
    
    def is_valid_move(self, col: int):
        return col >= 0 and col < self.columns and self.board[0][col] == '_'
    
    def make_move(self, col: int) -> bool:
        if not self.is_valid_move(col):
            return False
        
        r = self.__get_available_row(col)
        self.board[r][col] = self.currentPlayer
        return self.__check_win(r, col)
    
    def is_board_full(self) -> bool:
        for cell in self.board[0]:
            if cell == '_':
                return False
        return True
    
    def switch_player(self):
        self.currentPlayer =  'O' if self.currentPlayer == 'X' else 'X'
    
    def get_current_player(self):
        return self.currentPlayer
    
if __name__ == '__main__':
    game = Connect4(6, 7)
    print("Welcome to Connect 4!")
    game.display_board()

    while True:
        try:
            col = int(input(f"Player {game.get_current_player()}, enter your move (0-6): "))
        except ValueError:
            print("Invalid input. Please enter a number.")
            continue

        if not game.is_valid_move(col):
            print("Invalid move. Try again.")
            continue

        win = game.make_move(col)
        game.display_board()

        if win:
            print(f'Player {game.get_current_player()} wins the game!')
            break

        if game.is_board_full():
            print("Its a draw!")
            break

        game.switch_player()

File: glean/connect4.cpp
#include<bits/stdc++.h>
#include<iostream>
#include <iomanip>
using namespace std;

class Connect4 {
private:
    int rows, cols;
    vector<vector<char>> board;
    char currentPlayer;

    int countDirection(int row, int col, int deltaRow, int deltaCol) {
        int count = 0;
        int r = row + deltaRow;
        int c = col + deltaCol;
        while (r >= 0 && r < rows && c >= 0 && c < cols && board[r][c] == currentPlayer) {
            count++;
            r += deltaRow;
            c += deltaCol;
        }
        return count;
    }

public:
    Connect4(int n = 6, int m = 7) : rows(n), cols(m), board(n, vector<char>(m, '.')), currentPlayer('X') {}

    void displayBoard() {
        for (const auto& row : board) {
            for (char cell : row) {
                cout << setw(2) << cell << " ";
            }
            cout << endl;
        }
        for (int i = 0; i < cols; i++) {
            cout << setw(2) << i << " ";
        }
        cout << endl;
    }

    void resetBoard() {
        for (auto& row : board) {
            fill(row.begin(), row.end(), '.');
        }
        currentPlayer = 'X';
    }

    bool makeMove(int col) {
        if (!isValidMove(col)) return false;

        int row = getAvailableRow(col);
        board[row][col] = currentPlayer;
        
        return checkWin(row, col);
    }

    int getAvailableRow(int col) {
        for (int i = rows - 1; i >= 0; i--) {
            if (board[i][col] == '.') return i;
        }
        return -1;
    }

    bool isValidMove(int col) {
        return col >= 0 && col < cols && board[0][col] == '.';
    }

    void switchPlayer() {
        currentPlayer = (currentPlayer == 'X') ? 'O' : 'X';
    }

    bool checkWin(int row, int col) {
        // Horizontal
        if (1 + countDirection(row, col, 0, -1) + countDirection(row, col, 0, 1) >= 4)
            return true;
        // Vertical
        if (1 + countDirection(row, col, -1, 0) + countDirection(row, col, 1, 0) >= 4)
            return true;
        // Diagonal top right to bottom left
        if (1 + countDirection(row, col, -1, 1) + countDirection(row, col, 1, -1) >= 4)
            return true;
        // Diagonal top left to bottom right
        if (1 + countDirection(row, col, -1, -1) + countDirection(row, col, 1, 1) >= 4)
            return true;

        return false;
    }

    bool isBoardFull() {
        for (int i = 0; i < cols; i++) {
            if (board[0][i] == '.') return false;
        }
        return true;
    }

    char getCurrentPlayer() {
        return currentPlayer;
    }
};

int main() {
    Connect4* game = new Connect4(6, 7);
    bool gameWon = false;

    cout << "Welcome to Connect 4!\n";
    game->displayBoard();

    while (true) {
        int col;
        cout << "Player " << game->getCurrentPlayer() << ", enter your move (0-" << 6 << "): ";
        cin >> col;

        if (!cin || !game->isValidMove(col)) {
            cin.clear(); // Clear error flags if invalid input
            cin.ignore(1000, '\n'); // Discard bad input
            cout << "Invalid move. Try again.\n";
            continue;
        }

        bool win = game->makeMove(col);
        game->displayBoard();

        if (win) {
            cout << "Player " << game->getCurrentPlayer() << " wins!\n";
            break;
        }

        if (game->isBoardFull()) {
            cout << "It's a draw!\n";
            break;
        }

        game->switchPlayer();
    }
}